File: ts/src/storage/sqlite-wal.ts
import Database from "better-sqlite3";

// better-sqlite3's default busy timeout, which every caller opens its database with.
const WAL_RETRY_TIMEOUT_MS = 5_000;
const WAL_RETRY_POLL_MS = 50;
const WAL_RETRY_WAIT_BUFFER = new Int32Array(new SharedArrayBuffer(4));
// In-memory databases cannot use WAL; SQLite keeps reporting "memory" for them.
const ACCEPTED_JOURNAL_MODES = new Set(["wal", "memory"]);

/**
 * Switch the database to WAL, tolerating a concurrent writer.
 *
 * Switching a fresh database to WAL upgrades a read lock to a write lock, and
 * SQLite reports contention there as SQLITE_BUSY without consulting the busy
 * handler, so the `timeout` database option does not cover it.
 */
export function enableWal(db: Pick<Database.Database, "pragma">): void {
  const deadline = Date.now() + WAL_RETRY_TIMEOUT_MS;
  for (;;) {
    let mode: string | undefined;
    try {
      mode = String(db.pragma("journal_mode = WAL", { simple: true })).toLowerCase();
    } catch (error: unknown) {
      if (!(error instanceof Database.SqliteError) || error.code !== "SQLITE_BUSY") {
        throw error;
      }
      if (Date.now() >= deadline) {
        throw error;
      }
    }
    if (mode !== undefined) {
      if (ACCEPTED_JOURNAL_MODES.has(mode)) {
        return;
      }
      if (Date.now() >= deadline) {
        throw new Error(`could not switch to WAL journal mode; database reports '${mode}'`);
      }
    }
    Atomics.wait(WAL_RETRY_WAIT_BUFFER, 0, 0, WAL_RETRY_POLL_MS);
  }
}

File: autocontext/src/autocontext/storage/sqlite_wal.py
"""Switch SQLite databases to WAL journal mode under concurrent startup."""

from __future__ import annotations

import sqlite3
import time

# sqlite3.connect()'s default busy timeout, for connections opened without one.
DEFAULT_TIMEOUT_SECONDS = 5.0
_MAX_RETRY_DELAY_SECONDS = 0.05
# In-memory databases cannot use WAL; SQLite keeps reporting "memory" for them.
_ACCEPTED_JOURNAL_MODES = frozenset({"wal", "memory"})


def enable_wal(connection: sqlite3.Connection, *, timeout_seconds: float = DEFAULT_TIMEOUT_SECONDS) -> None:
    """Switch the database to WAL, tolerating a concurrent writer.

    The conversion needs a brief exclusive lock, and unlike ordinary statements
    `PRAGMA journal_mode` returns SQLITE_BUSY *without* invoking the busy
    handler, so neither `sqlite3.connect(timeout=...)` nor `PRAGMA busy_timeout`
    covers it. Retry here for up to ``timeout_seconds``, which callers set to
    their connection's busy timeout. Losing the race is a success rather than
    an error: the journal mode lives in the database header, so once any
    initializer wins, every later connection already sees WAL.
    """
    deadline = time.monotonic() + timeout_seconds
    delay = 0.001
    while True:
        try:
            row = connection.execute("PRAGMA journal_mode=WAL").fetchone()
        except sqlite3.OperationalError as error:
            message = str(error).lower()
            if ("locked" not in message and "busy" not in message) or time.monotonic() >= deadline:
                raise
        else:
            mode = str(row[0]).lower() if row is not None else ""
            if mode in _ACCEPTED_JOURNAL_MODES:
                return
            if time.monotonic() >= deadline:
                raise sqlite3.OperationalError(f"could not switch to WAL journal mode; database reports {mode!r}")
        time.sleep(delay)
        delay = min(delay * 2, _MAX_RETRY_DELAY_SECONDS)
